File: 2023/02/main.py
#!/usr/bin/env python3
from math import prod
from sys import argv

# don't close files, it's ok ;0_0;

# first

print("1)", sum(map(lambda v: v[0], filter(
    lambda v: v[1],
    [(i, all(all((int(cubes.split()[0]) <=
                  {"red": 12, "green": 13, "blue": 14}[cubes.split()[1]])
                 for cubes in game.split(", "))
             for game in line.strip().split(": ")[1].split("; ")))
     for i, line in enumerate(open(argv[1], "r").readlines(), start=1)]
))))

# second

res2 = 0
for line in (line.strip() for line in open(argv[1], "r").readlines()):
    res = {"red": 0, "green": 0, "blue": 0}
    cubes = [[cube.split() for cube in game.split(", ")]
             for game in line.strip().split(": ")[1].split("; ")]
    for cube in cubes:
        for v, k in cube:
            if res[k] < (v := int(v)):
                res[k] = v
    res2 += prod(map(lambda v: v[1], res.items()))
print("2)", res2)

File: 2023/02/main.rs
use std::{
    collections::HashMap,
    env::args,
    fs::File,
    io::{BufRead, BufReader},
};

fn main() {
    let br = BufReader::new(File::open(args().collect::<Vec<_>>().get(1).unwrap()).unwrap());
    let max_values: HashMap<&str, u64> = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);

    let mut res1 = 0u64;
    let mut res2 = 0u64;

    for (li, line) in br.lines().enumerate() {
        match line {
            Ok(line) => {
                let mut possible = true;

                'first: for game in line.split(": ").last().unwrap().split("; ") {
                    for cubes in game.split(", ") {
                        let mut cubsplit = cubes.split(' ');
                        let (amount, color): (u64, &str) = (
                            cubsplit.next().unwrap().parse().unwrap(),
                            cubsplit.next().unwrap(),
                        );
                        if amount > max_values[color] {
                            possible = false;
                            break 'first;
                        }
                    }
                }

                if possible {
                    res1 += li as u64 + 1;
                }

                let mut mins: HashMap<&str, u64> = HashMap::from([("red", 0), ("green", 0), ("blue", 0)]);

                for game in line.split(": ").last().unwrap().split("; ") {
                    for cubes in game.split(", ") {
                        let mut cubsplit = cubes.split(' ');
                        let (amount, color): (u64, &str) = (
                            cubsplit.next().unwrap().parse().unwrap(),
                            cubsplit.next().unwrap(),
                        );
                        if mins[color] < amount {
                            mins.insert(color, amount);
                        }
                    }
                }

                res2 += mins.into_iter().reduce(|(k, v1), (.., v2)| (k, v1 * v2)).unwrap().1;
            }
            Err(_) => break,
        }
    }

    println!("1) {res1}");
    println!("2) {res2}");
}
